modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/LocalDatabases.java

```diff
@@ -0,0 +1,208 @@
+            defaultDatabases.keySet(), configDatabases.keySet(), geoipConfigDir);
+    }
+
+    DatabaseReaderLazyLoader getDatabase(String name) {
+        DatabaseReaderLazyLoader database = configDatabases.get(name);
+        if (database != null) {
+            return database;
+        } else {
+            return defaultDatabases.get(name);
+        }
+    }
+
+    List<DatabaseReaderLazyLoader> getAllDatabases() {
+        List<DatabaseReaderLazyLoader> all = new ArrayList<>(defaultDatabases.values());
+        all.addAll(configDatabases.values());
+        return all;
+    }
+
+    Map<String, DatabaseReaderLazyLoader> getDefaultDatabases() {
+        return defaultDatabases;
+    }
+
+    Map<String, DatabaseReaderLazyLoader> getConfigDatabases() {
+        return configDatabases;
+    }
+
+    void updateDatabase(Path file, boolean update) {
+        String databaseFileName = file.getFileName().toString();
+        try {
+            Map<String, DatabaseReaderLazyLoader> databases = new HashMap<>(this.configDatabases);
+            if (update) {
+                LOGGER.info("database file changed [{}], reload database...", file);
+                DatabaseReaderLazyLoader loader = new DatabaseReaderLazyLoader(cache, file);
+                DatabaseReaderLazyLoader existing = databases.put(databaseFileName, loader);
+                if (existing != null) {
+                    existing.close();
+                    int numEntriesEvicted = cache.purgeCacheEntriesForDatabase(file);
+                    LOGGER.info("evicted [{}] entries from cache after reloading database [{}]", numEntriesEvicted, file);
+                }
+            } else {
+                LOGGER.info("database file removed [{}], close database...", file);
+                DatabaseReaderLazyLoader existing = databases.remove(databaseFileName);
+                assert existing != null;
+                existing.close();
+                int numEntriesEvicted = cache.purgeCacheEntriesForDatabase(file);
+                LOGGER.info("evicted [{}] entries from cache after reloading database [{}]", numEntriesEvicted, file);
+            }
+            this.configDatabases = Map.copyOf(databases);
+        } catch (Exception e) {
+            LOGGER.error((Supplier<?>) () -> new ParameterizedMessage("failed to update database [{}]", databaseFileName), e);
+        }
+    }
+
+    Map<String, DatabaseReaderLazyLoader> initDefaultDatabases(Path geoipModuleDir) {
+        Map<String, DatabaseReaderLazyLoader> databases = new HashMap<>();
+
+        for (String filename : DEFAULT_DATABASE_FILENAMES) {
+            Path source = geoipModuleDir.resolve(filename);
+            assert Files.exists(source);
+            String databaseFileName = source.getFileName().toString();
+            DatabaseReaderLazyLoader loader = new DatabaseReaderLazyLoader(cache, source);
+            databases.put(databaseFileName, loader);
+        }
+
+        return Map.copyOf(databases);
```

Comment:
Any reason for creating copy here? `Collections.unmodifiableMap(databases)` should suffice I think

Reply:
It only needs to be immutable, so like you suggest `Collections.unmodifiableMap(...)` suffices.